### PR TITLE
Fix lease acquisition.

### LINKED ISSUE #3050: lease acquisition uses non-idempotent txn closure, probably breaking OSC
apply this diff:

```diff
diff --git a/server/testserver.go b/server/testserver.go
index 2fad316..5e7f5ea 100644
--- a/server/testserver.go
+++ b/server/testserver.go
@@ -18,6 +18,8 @@
 package server

 import (
+       "time"
+
        "github.com/cockroachdb/cockroach/client"
        "github.com/cockroachdb/cockroach/config"
        "github.com/cockroachdb/cockroach/gossip"
@@ -62,7 +64,7 @@ func NewTestContext() *Context {
        // MaxOffset is the maximum offset for clocks in the cluster.
        // This is mostly irrelevant except when testing reads within
        // uncertainty intervals.
-       ctx.MaxOffset = 0
+       ctx.MaxOffset = 50 * time.Millisecond

        // Load test certs. In addition, the tests requiring certs
        // need to call security.SetReadFileFn(securitytest.Asset)
```

It leads to some transaction restarts (which would also happen in production).
Then,

```
--- FAIL: TestLeaseManager (0.03s)
        lease_test.go:90: expected /1/1, but found
--- FAIL: TestLeaseManagerReacquire (0.02s)
        lease_test.go:253: expected refcount of 2, but found 3
```

The issue is that `(*LeaseManager).Acquire` takes a transaction, but changes its internal state according to operations carried out within that transaction without ascertaining that the transaction actually commits (which, of course, it may not).

The resulting state looks quite inconsistent, with refcount overcounting being on the harmless side. In the first test failure, it appears that the node thinks it has a lease while nothing's been persisted to the lease table, breaking online schema changes.

### PR DESCRIPTION
**Pull request by petermattis**
Inserting an entry into the lease table now uses a separate transaction
from the transaction passed to LeaseStore.Acquire(). We need to use the
supplied (user) transaction to look up the descriptor because the
descriptor might have been created with the transaction. We need to use
a separate transaction to create the lease entry because the
supplied (user) transaction may be aborted or retried and the result
from LeaseStore.Acquire() is stored in an in-memory map which cannot
handle aborts/retries.

Fixes #3050.

[Review on Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/3081)

### CODE REVIEW

**Comment by bdarnell** [comment]
LGTM

**Comment by tamird** [comment]
Reviewed 2 of 2 files at r1.
Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

<sup>**[sql/lease_test.go, line 256 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3081#-K2mDdtwoFFozHUJLd2B)** ([raw file](https://github.com/cockroachdb/cockroach/blob/80519081349643aecf9f7d3d43ae0033d587e6a0/sql/lease_test.go#L256)):</sup>
How come the old way of manipulating the clock no longer suffices?


---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3081)

**Comment by petermattis** [comment]
Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

<sup>**[sql/lease_test.go, line 256 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3081#-K2mDdtwoFFozHUJLd2B)** ([raw file](https://github.com/cockroachdb/cockroach/blob/80519081349643aecf9f7d3d43ae0033d587e6a0/sql/lease_test.go#L256)):</sup>
I'm not sure. `Store.Send` was complaining about `"Rejecting command with timestamp in the future"`. Perhaps this never worked correctly.


---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3081)

**Comment by tamird** [comment]
Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

<sup>**[sql/lease_test.go, line 256 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3081#-K2mDdtwoFFozHUJLd2B)** ([raw file](https://github.com/cockroachdb/cockroach/blob/80519081349643aecf9f7d3d43ae0033d587e6a0/sql/lease_test.go#L256)):</sup>
Hm, that's unsatisfying. 


---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3081)

File: sql/lease.go [hunk 1]
```diff
@@ -39,12 +39,14 @@ import (
...
-)
-
 var (
+	// LeaseDuration is the mean duration a lease will be acquired for. The
+	// actual duration is jittered in the range
+	// [0.75,1.25]*LeaseDuration. Exported for testing purposes only.
+	LeaseDuration = 5 * time.Minute
+	// MinLeaseDuration is the minimum duration a lease will have remaining upon
+	// acquisition. Exported for testing purposes only.
+	MinLeaseDuration       = time.Minute
```
Comment (vivekmenezes) [inline sql/lease.go]: wanna add the Test prefix

**Comment by petermattis** [comment]
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

<sup>**[sql/lease_test.go, line 256 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3081#-K2mDdtwoFFozHUJLd2B)** ([raw file](https://github.com/cockroachdb/cockroach/blob/80519081349643aecf9f7d3d43ae0033d587e6a0/sql/lease_test.go#L256)):</sup>
Agreed. I asked @tschottdorf at lunch at he said that adjusting the logical clock is not kosher. Not sure how this was ever working. If I want to adjust the clock I need to create a manual-clock that is wrapped by an `hlc.Clock` and adjust that. I'll look into doing so this afternoon.


---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3081)

**Comment by vivekmenezes** [comment]
Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

<sup>**[sql/lease.go, line 137 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3081#-K2mbXF93I2kCnIzyPFT)** ([raw file](https://github.com/cockroachdb/cockroach/blob/14052518b6baddc0c7b71ac3e0972eb52b03df43/sql/lease.go#L137)):</sup>
nice explanation. Would be valuable to also add

BEGIN
CREATE TABLE kv (k CHAR PRIMARY, v CHAR)
INSERT INTO TABLE kv VALUES ('a', 'b')
END

The lease acquisition happens in the INSERT and the entire transaction above fails resulting in the dangling lease.


---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3081)

**Comment by tschottdorf** [comment]
LGTM.


---

Reviewed 2 of 2 files at r1.
Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

<sup>**[sql/lease_test.go, line 256 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3081#-K2mDdtwoFFozHUJLd2B)** ([raw file](https://github.com/cockroachdb/cockroach/blob/80519081349643aecf9f7d3d43ae0033d587e6a0/sql/lease_test.go#L256)):</sup>
@tamird here's some satisfaction: the change which started skipping this test is the one that actually set a nontrivial `MaxOffset` on the clock. Now look at the offending code:

```go
		if s.Clock().MaxOffset() > 0 {
			// Once a command is submitted to raft, all replicas' logical
			// clocks will be ratcheted forward to match. If the command
			// appears to come from a node with a bad clock, reject it now
			// before we reach that point.
			offset := time.Duration(ba.Timestamp.WallTime - s.Clock().PhysicalNow())
			if offset > s.Clock().MaxOffset() {
				return nil, roachpb.NewError(util.Errorf("Rejecting command with timestamp in the future: %d (%s ahead)",
					ba.Timestamp.WallTime, offset))
			}
```

so that's why it worked before.


---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3081)

**Comment by tamird** [comment]
Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

<sup>**[sql/lease_test.go, line 256 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3081#-K2mDdtwoFFozHUJLd2B)** ([raw file](https://github.com/cockroachdb/cockroach/blob/80519081349643aecf9f7d3d43ae0033d587e6a0/sql/lease_test.go#L256)):</sup>
Ah, so since this test has access to the clock...it could just `SetMaxOffset` to zero here?


---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3081)

**Comment by tschottdorf** [comment]
Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

<sup>**[sql/lease_test.go, line 256 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3081#-K2mDdtwoFFozHUJLd2B)** ([raw file](https://github.com/cockroachdb/cockroach/blob/80519081349643aecf9f7d3d43ae0033d587e6a0/sql/lease_test.go#L256)):</sup>
suppose it could, though that seems dirty to me.


---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3081)

**Comment by tamird** [comment]
Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

<sup>**[sql/lease_test.go, line 256 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3081#-K2mDdtwoFFozHUJLd2B)** ([raw file](https://github.com/cockroachdb/cockroach/blob/80519081349643aecf9f7d3d43ae0033d587e6a0/sql/lease_test.go#L256)):</sup>
Yep, I agree.


---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3081)

**Pull request merged**